src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs

```diff
@@ -527,6 +527,19 @@ protected override Expression VisitFormattedExpression(FormattedBindingParserNod
 
         protected override Expression VisitVoid(VoidBindingParserNode node) => Expression.Default(typeof(void));
 
+        protected override Expression VisitArrayInitializer(ArrayInitializerExpression node)
+        {
+            var initializers = node.ElementInitializers.Select(e => Visit(e));
+
+            var firstInitializer = initializers.FirstOrDefault();
+
+            var firstElementType = firstInitializer?.Type ?? typeof(object);
+
+            var arrayElementType = initializers.All(i => i.Type.IsAssignableFrom(firstElementType)) ? firstElementType : typeof(object);
+
+            return Expression.NewArrayInit(arrayElementType, initializers.Select(i => Expression.Convert(i, arrayElementType)));
```

Comment:
Maybe using `TypeConversions.ImplicitConversion()` instead of Expression.Convert and IsAssignableTo

Reply:
I have tried `TypeConversions.ImplicitConversion` at first but it is intended for html attribute values and does such random magic as to be unhelpful. The numeric conversions are contraproductive here for example.